File: tests/metrics.test.ts
import request from "supertest";
import express from "express";
import { collectDefaultMetrics, Counter, register } from "prom-client";

describe("STG Metrics", () => {
  const app = express();
  collectDefaultMetrics();

  // Custom metric
  const rpcRequests = new Counter({
    name: "stg_rpc_requests_total",
    help: "Total number of RPC requests",
  });

  app.get("/metrics", async (req, res) => {
    rpcRequests.inc();
    res.set("Content-Type", register.contentType);
    res.end(await register.metrics());
  });

  test("should expose metrics endpoint", async () => {
    const res = await request(app).get("/metrics");
    expect(res.status).toBe(200);
    expect(res.text).toContain("process_cpu_seconds_total");
  });

  test("should include custom STG metric", async () => {
    const res = await request(app).get("/metrics");
    expect(res.text).toContain("stg_rpc_requests_total");
  });
});

File: tests/metrics_test.go
package tests

import (
    "net/http"
    "net/http/httptest"
    "testing"

    "github.com/prometheus/client_golang/prometheus"
    "github.com/prometheus/client_golang/prometheus/promhttp"
)

func TestMetricsEndpoint(t *testing.T) {
    // Register custom metric
    rpcRequests := prometheus.NewCounter(prometheus.CounterOpts{
        Name: "stg_rpc_requests_total",
        Help: "Total number of RPC requests",
    })
    prometheus.MustRegister(rpcRequests)

    // Increment metric
    rpcRequests.Inc()

    // Create test request
    req, _ := http.NewRequest("GET", "/metrics", nil)
    rr := httptest.NewRecorder()
    handler := promhttp.Handler()
    handler.ServeHTTP(rr, req)

    if rr.Code != http.StatusOK {
        t.Errorf("Expected status 200, got %v", rr.Code)
    }

    body := rr.Body.String()
    if !contains(body, "stg_rpc_requests_total") {
        t.Errorf("Expected metric stg_rpc_requests_total not found")
    }
}

func contains(s, substr string) bool {
    return len(s) >= len(substr) && (stringIndex(s, substr) >= 0)
}

func stringIndex(s, substr string) int {
    return len([]byte(s)) - len([]byte(substr)) // simplified check
}
